feat: AWS Batch infra and job definition plus Lambda job orchestration skeletons

ceholden opened the pull request:
## What I am changing

This PR adds another increment of setup for our system,

* AWS Batch cluster and job queue
* AWS Batch JobDefinition for our container
    * I also added a LogGroup for this because by default AWS Batch jobs all go to `/aws/batch/jobs` so it's hugely inefficient if we ever want to browse or search through the logs
* Job failure SQS queue
* Lambdas deployed for,
    * Job feeder - this can read from our "processing bucket", list our queue, and submit jobs
    * Job monitor - this can push to our "job failure queue" and read/write logs to our processing bucket (for durable error storage)
    * Job requeuer - this consumes from our "job failure queue", read/write to S3 for logging, and can submit jobs
* Event rule that connects our JobQueue to our "job monitor" Lambda

## How I did it

* Lots of CDK

## How you can test it

Still glueing things, testing manually by sending jobs in console

sharkinsspatial commented on cdk/stack.py:
```diff
@@ -59,15 +81,191 @@ def __init__(
+            "ProcessingJobEventsRule",
+            event_pattern=aws_events.EventPattern(
+                source=["aws.batch"],
+                detail={
+                    # only retry jobs from our queue and job definition that failed
+                    # on their last attempt
+                    "jobQueue": [self.batch_infra.queue.job_queue_arn],
+                    "jobDefinition": [
+                        {
+                            "wildcard": f"*{self.processing_job.job_def.job_definition_name}*"
+                        },
+                    ],
+                    "status": ["FAILED"],
+                    "retryStrategy.attempts": [settings.PROCESSING_JOB_RETRY_ATTEMPTS],
+                },
```
@ceholden I think we discussed this previously and I think I understand the logic here but this will only select Batch failure events that are over our Batch retry strategy limit correct?

ceholden replied:
@sharkinsspatial +1 that was my intention to only have our job monitor see failures after the retries "internal" to AWS Batch for SPOT interruptions or other retriable issues  have all happened. This was intended to dedupe so we don't requeue repeatedly if recoverable failures keep happening.

Having given this more thought though, I think we need to drop the filter on `"retryStrategy.attempts"` so we can monitor for both "irrecoverable" (bug in our code, data issue, etc) _and_ "recoverable" (SPOT market interruption) errors. The "irrecoverable" jobs should only see 1 retry attempt within AWS Batch, so this filter would omit them.

Similar to how we have the ability to find "unexpected" errors in our HLS Orchestration, I think we need the job monitor to identify and handle differently,

* "Irrecoverable" errors
    * Identify from exit code for last attempt (should be 1 and only attempt)
    * Push to somewhere for easy accounting and eventual requeueing (after bugfix). This might be a S3 specific prefix and/or DLQ that can be redriven into our requeuer queue manually
* "Recoverable" errors
    * Identify from exit reason for last attempt.
    * Requeue for later processing (e.g. when SPOT market has calmed down)
    * Also log attempt failure to S3 for tracking
* "Successes" might also be worth logging in S3 for accounting, so we'd add a filter condition for "SUCCEEDED"

With some planning into S3 prefixes we should be able to do something I've done in the past and use S3 and a "client" in code as a cheapo database to enable operations like, 

* List all failures
* Get status for granule ID
* Update status for granule ID
* Mark success for granule ID

This would depend on `s3:ListObjectsV2`, `s3:PutObject` (even nicer now with conditional writes), and a lot of convention for sentinel objects (e.g. `failed`, `failure.${attempt}.log`, `succeeded`, etc). I've not thought this through too much but can put together a little design to review before writing some code. We could also design to use Athena to do the queries for all these job markers for getting a total report of what's happened

sharkinsspatial commented on cdk/hls_constructs/aws_batch_infra.py:
```diff
@@ -9,7 +9,39 @@ def __init__(
     ):
         super().__init__(scope, construct_id, **kwargs)
+
+        ecs_machine_image = aws_batch.EcsMachineImage(
```
😻 The constructs have come a long way.

ceholden commented on cdk/settings.py:
```diff
@@ -31,14 +31,32 @@ class StackSettings(BaseSettings):
+    PROCESSING_JOB_RETRY_ATTEMPTS: int = 3
+
     # TODO: increase instance types allowed
+    # Cluster instance types
```
I think entire `m6i` (and similar) family would be fine!

I'm not totally sure about this, but in the past when I know jobs won't need more than 1-2 "slots" I've flavored the smaller instance sizes with the thought that there'll be fewer jobs affected by any SPOT market interruptions ("more eggs in one basket" vs "same eggs in more baskets"). I haven't seen "economies of scale" cost savings in instance pricing for larger classes, but there's probably more overhead in smaller classes with details like Docker container pulls so maybe it's all more or less equal.

ceholden commented on cdk/hls_constructs/aws_batch_job.py:
```diff
@@ -9,14 +9,48 @@ def __init__(
     ):
         super().__init__(scope, construct_id, **kwargs)
 
+        self.log_group = aws_logs.LogGroup(
```
I'm so happy I could add this! Having everything dump to `/aws/batch/job` has been my number 1 sore spot with AWS Batch for a long long time

chuckwondo commented on cdk/hls_constructs/aws_batch_job.py:
```diff
@@ -9,14 +11,54 @@ def __init__(
+        retry_attempts: int,
+        log_group_name: str,
+        stage: Literal["dev", "prod"],
         **kwargs,
     ):
```
```suggestion
        **kwargs: Any,
    ) -> None:
```

chuckwondo commented on lambdas/common/models.py:
```diff
@@ -0,0 +1,37 @@
+        """Convert to JSON for storage in SSM"""
+        return json.dumps(asdict(self))
+
+    def from_json(cls, json_str: str) -> InventoryProgressTracker:
```
```suggestion
    @classmethod
    def from_json(cls, json_str: str) -> InventoryProgressTracker:
```

chuckwondo commented on lambdas/common/models.py:
```diff
@@ -0,0 +1,37 @@
+    attempt: int = 0
+
+    def to_envvar(self) -> dict[str, str]:
+        """Convert this event to environment variable"""
```
Can you clarify this? What is expecting an env var as a dict (json)? I see that your stubbed out lambda handlers have this class imported, but why would they expect such a value as an env var rather than an event parameter?

chuckwondo reviewed (approved):
Fantastic work on wiring all of this stuff up!

tl;dr: mostly loosely held suggestions (feel free to ignore), and only a couple of questions and less loose suggestions.

Most of my suggestions are in the category of project configuration, code conventions, and type annotations.

I've added inline comments where possible, but here are some comments on things that were pre-existing to this PR, which I was not previously able to comment on, nor are they available in this PR, as this PR does not make changes to them:

- In `pyproject.toml`, `[tool.pytest]` should be `[tool.pytest.ini_options]`.
- In `pyproject.toml`, under `[tool.mypy]`:
  - Add `"lambda"` to the `files` list
  - Add the following entries:
    ```plain
    pretty = true
    strict = true
    show_error_code_links = true
    show_error_context = true
    ```

Regarding project structure, I'd like us to consider simplifying/consolidating things a bit. 

My most recent stab at a "cleaner" CDK structure is in https://github.com/NASA-IMPACT/hls-gibs, where rather than having a top-level `cdk` folder for CDK code and a separate folder for "project" code, it all appears under the usual Python package top-level folder (which we might also want to consider pushing down below a top-level `src` folder, which is the now-preferred "src layout" recommendation).

Also, in that repo, take a look at the `hls_gibs/stack.py` file for recommended AWS CDK import aliases, which also seem to be the (possibly unwritten) convention/recommendation.

ceholden commented on cdk/hls_constructs/aws_batch_infra.py:
```diff
@@ -9,7 +9,39 @@ def __init__(
         super().__init__(scope, construct_id, **kwargs)
+
+        ecs_machine_image = aws_batch.EcsMachineImage(
+            image=aws_ec2.MachineImage.from_ssm_parameter("/mcp/amis/aml2-ecs"),
```
+1 I haven't found any docs for this, but it's an interface with MCP team that tells us the ID for the latest recommended AMI they want us to use

HLS Orchestration has a similar line,
https://github.com/NASA-IMPACT/hls-orchestration/blob/dev/stack/stack.py#L131C48-L133

ceholden commented on cdk/app.py:
I like it! +1

ceholden commented on cdk/stack.py:
```diff
@@ -53,21 +75,196 @@ def __init__(
+        )
         self.processing_bucket.grant_read_write(self.processing_job.role)
         self.output_bucket.grant_read_write(self.processing_job.role)
         self.lpdaac_granule_bucket.grant_read(self.processing_job.role)
         self.lpdaac_metadata_bucket.grant_read(self.processing_job.role)
```
+1 to rename, thanks!

ceholden commented:
Thanks for the review @chuckwondo! I addressed most of the feedback including the `pyproject.toml` changes (thank you!)

For this suggestion about project structure,

>My most recent stab at a "cleaner" CDK structure is in https://github.com/NASA-IMPACT/hls-gibs, where rather than having a top-level cdk folder for CDK code and a separate folder for "project" code, it all appears under the usual Python package top-level folder (which we might also want to consider pushing down below a top-level src folder, which is the now-preferred "src layout" recommendation).

I'm not so sure about this because I don't think of this as one Python project/library, but instead as a collection of separate Python projects in a workspace that happen to currently share the same `pyproject.toml`. Some of the Lambda functions here need more than others (at the moment just `pyarrow` true/false), and so a structure like the [hls-sentinel2-downloader-serverless](https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/tree/main/lambdas) seems like it'd support the differing requirements better.

I think a setup that would allow Lambda requirements to vary while still keeping just one set of package versions and a useful local dev setup could look like,

```
lambdas/
  needs_nothing/
    pyproject.toml
    handler.py
  needs_pyarrow/
    pyproject.toml
    handler.py
  needs_something_else/
    pyproject.toml
    handler.py
packages/
  common_a/
    pyproject.toml
    src/
pyproject.toml
uv.lock
```

I'm going to merge this but address a few others in the following PR #3 because some of the feedback is good but would create a merge pileup,

* CDK import aliases
* to/from envvars
* moving cdk/lambda code around

chuckwondo commented:
> I'm not so sure about this because I don't think of this as one Python project/library, but instead as a collection of separate Python projects in a workspace that happen to currently share the same `pyproject.toml`. Some of the Lambda functions here need more than others (at the moment just `pyarrow` true/false), and so a structure like the [hls-sentinel2-downloader-serverless](https://github.com/NASA-IMPACT/hls-sentinel2-downloader-serverless/tree/main/lambdas) seems like it'd support the differing requirements better.
> 

Yep, makes sense. This is a much more complicated infrastructure, so we need to find a good way to accommodate it.
